### ISSUE TITLE
History Retrieval

### ISSUE THREAD

**Issue by jaredbroad** [QuantConnect/Lean#21]
Implement a system for retrieving historical data as an array. The QCAlgorithm helper methods should match the interface below:

Get all data from start to end; for specified symbol:
`IEnumerable<BaseData> History(string symbol, DateTime start, DateTime end);`

Overload: Get all data from *now* back a timspan period for symbol and return it as an enumerable list:
`IEnumerable<BaseData> History(string symbol, TimeSpan period);`

QCAlgorithm should have an IHistoryManager plugin point which is set by Lean Engine according to the DataFeed. The IDataFeed would provide a GetHistoryManager method which returns a IHistoryManager, and that is set in the algorithm.

IHistoryManager implementation for the FileSystemDataFeed would implement this with a physical pull from the file system for the associated dates.

**Comment by jaredbroad** [comment]
Feedback from a user: This is better implemented in a "Warm Up" style -- where the user actually requests a fixed period of data before the start date and the algorithm automatically warms up / pre-populates all the arrays as it normally would.

This makes the warm up process inline with normal algorithm code: e.g.
```
SetStartDate(2014, 1, 1);
SetWarmUp(TimeSpan.FromDays(14));
```

With this example data would be pumped in from Dec 15th, but orders would only be permitted after the 1st (or once the warm up period is completed in case of live trading).

**Comment by jaredbroad** [comment]
Closed in 4ae1f4f5b61
